manifest.go

```diff
@@ -1327,6 +1344,40 @@ func (m *ManifestListWriter) AddManifests(files []ManifestFile) error {
 				return err
 			}
 		}
+	case 3:
+		for _, file := range files {
+			if file.Version() != 3 {
+				return fmt.Errorf("%w: ManifestListWriter only supports version 3 manifest files", ErrInvalidArgument)
+			}
+			wrapped := *(file.(*manifestFile))
+
+			// Ref: https://github.com/apache/iceberg/blob/ea2071568dc66148b483a82eefedcd2992b435f7/core/src/main/java/org/apache/iceberg/ManifestListWriter.java#L157-L168
+			if wrapped.Content == ManifestContentData && wrapped.FirstRowId == nil {
+				if m.nextRowID != nil {
+					wrapped.FirstRowId = m.nextRowID
+					*m.nextRowID += wrapped.ExistingRowsCount + wrapped.AddedRowsCount
+				}
+			}
+
+			// Ref: https://github.com/apache/iceberg/blob/ea2071568dc66148b483a82eefedcd2992b435f7/core/src/main/java/org/apache/iceberg/V3Metadata.java#L98-L122
+			if wrapped.SeqNumber == -1 {
+				if m.commitSnapshotID != wrapped.AddedSnapshotID {
+					return fmt.Errorf("found unassigned sequence number for a manifest from snapshot %d != %d",
+						m.commitSnapshotID, wrapped.AddedSnapshotID)
+				}
+				wrapped.SeqNumber = m.sequenceNumber
+			}
+
+			if wrapped.MinSeqNumber == -1 {
+				if m.commitSnapshotID != wrapped.AddedSnapshotID {
+					return fmt.Errorf("found unassigned sequence number for a manifest from snapshot: %d", wrapped.AddedSnapshotID)
+				}
+				wrapped.MinSeqNumber = m.sequenceNumber
+			}
+			if err := m.writer.Encode(wrapped); err != nil {
+				return err
+			}
```

Comment:
since this logic is identical between v2 and v3, is there any way we can consolidate the logic and reduce the duplication?

Reply:
Initially, I want it to be separated from version 2 to avoid changing the existing logic as much as possible. But yes you are right, the only different is from the first row id for lineage tracking. I have combined both logic